domain/entity/book.go

```diff
@@ -14,3 +16,24 @@ type Book struct {
 	CreatedAt time.Time
 	UpdatedAt time.Time
 }
+
+//NewBook create a new book
+func NewBook(title string, author string, pages int, quantity int) (*Book, error) {
+	b := &Book{
+		ID:        NewID(),
+		Title:     title,
+		Author:    author,
+		Pages:     pages,
+		Quantity:  quantity,
+		CreatedAt: time.Now(),
+	}
+	return b, nil
+}
+
+//Validate validate book
+func (b *Book) Validate() error {
+	if b.Title == "" || b.Author == "" || b.Pages <= 0 {
```

Comment:
If you want to elaborate the DDD example beyond the clean architecture concept, you could consider adding some more domain types. Like Value Objects in this case maybe for `Title`. They would have their own validation rules, like maybe `maxLength`. On `User` you'd have a `Password` value object that checks password strength, according to a company's security requirements (that would make it a true Domain business rule).

Validation is always a bit tricky:
- Do you validate as close as possible to the outer world, e.g. in Infra endpoints (e.g. API)
- Do you validate in the domain layer, making it completely self-contained
- Do you validate at both locations. Stuff may be validated twice (boilerplate) or you separate on type of validation

The first option is a fail-early strategy that may save processing and possibly errors occurring deeper down in the architecture. But you may miss domain rules by moving all or most validation there. But you could ponder if a `GetUser` without a `UserID` really constitutes a domain rule (it would be part of the ubiquitous language that you discuss with the customer).